src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -1,64 +1,386 @@
+        val componentData: MutableList<Component> = mutableListOf()
+        var previousListSelection: JBList<*>? = null
+        val averagePreferredWidth = calculateAveragePreferredWidth(traces)
+
+        traces.forEach { trace ->
+            if (trace == null) {
+                componentData.add(Separator())
+                return@forEach
+            }
+
+            val vertex = createCoroutineTraceUI(trace, lastRunningStackFrame, averagePreferredWidth)
+            vertex.addListSelectionListener { e ->
+                val currentList = e.source as? JBList<*> ?: return@addListSelectionListener
+                if (previousListSelection != currentList) {
+                    previousListSelection?.clearSelection()
+                }
+                previousListSelection = currentList
+            }
+            componentData.add(vertex)
+        }
+
+        val forest = ContainerWithEdges()
+        componentData.forEach { forest.add(it) }
+        forest.layout = ForestLayout()
+
+        return JBScrollPane(forest)
+    }
+
+    private fun calculateAveragePreferredWidth(traces: List<CoroutineTrace?>): Int {
+        val validTraces = traces.filterNotNull()
+        val maxWidth = validTraces.sumOf { trace ->
+            val vertex = createVertexAndData(trace).first
+            vertex.preferredWidth
+        }
+        return maxWidth / validTraces.size
+    }
+
+    private fun createCoroutineTraces(rootValue: Node): List<CoroutineTrace?> {
+        val stack = Stack<Pair<Node, Int>>().apply { push(rootValue to 0) }
+        val parentStack = Stack<Node>()
+        var previousLevel: Int? = null
+        val coroutineTraces = mutableListOf<CoroutineTrace?>()
+
+        while (stack.isNotEmpty()) {
+            val (currentNode, currentLevel) = stack.pop()
+            val parent = if (parentStack.isNotEmpty()) parentStack.pop() else null
+
+            if (parent != null && parent.num != currentNode.num) {
+                val currentTrace = CoroutineTrace(
+                    mutableListOf(currentNode.stackFrameLocation),
+                    CoroutineStacksBundle.message("number.of.coroutines.active", currentNode.num),
+                    currentNode.coroutinesActive
+                )
+                repeat((previousLevel ?: 0) - currentLevel + 1) {
+                    coroutineTraces.add(null)
+                }
+                coroutineTraces.add(currentTrace)
+                previousLevel = currentLevel
+            } else if (parent != null) {
+                coroutineTraces.lastOrNull()?.locations?.add(0, currentNode.stackFrameLocation)
+            }
+
+            currentNode.children.values.reversed().forEach { child ->
+                stack.push(child to (currentLevel + 1))
+                parentStack.push(currentNode)
+            }
+        }
+
+        return coroutineTraces
+    }
+
+    private fun SuspendContextImpl.createCoroutineTraceUI(
```

Comment:
It would be better extract the coroutine trace UI to a separate class and handle all customisation there:

```
class CoroutineFramesList : JBList<String> {
 ...
}
```